File: backend/src/main/java/com/sdrouet/easy_restaurant/service/UserService.java
package com.sdrouet.easy_restaurant.service;

import com.sdrouet.easy_restaurant.dto.auth.RegisterUserRequest;
import com.sdrouet.easy_restaurant.dto.user.UpdateUserRequest;
import com.sdrouet.easy_restaurant.dto.user.UserResponse;
import com.sdrouet.easy_restaurant.entity.User;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface UserService {
    User register(RegisterUserRequest request);

    UserResponse updateUser(UpdateUserRequest updateUserRequest);

    User changeUserStatus(Long userId);

    User getUserById(Long userId);

    Page<User> getUsers(Pageable pageable, String username, String email, String name);
}

File: frontend/src/app/core/service/user-service.ts
import { inject, Injectable } from '@angular/core';
import { UserApiService } from '@data/api/user-api-service';
import { ApiErrorDetail } from '@data/interfaces/common-interface';
import { PageableParams } from '@data/interfaces/pageable-interface';
import { RegisterUser, User } from '@data/interfaces/user-interface';
import { catchError, throwError } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class UserService {
  private userApiService = inject(UserApiService);

  getUserById(userId: number) {
    return this.userApiService.getUserById(userId)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }

  updateUser(updateData: Partial<User>) {
    return this.userApiService.updateUser(updateData)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }

  registerUser(userData: RegisterUser) {
    return this.userApiService.registerUser(userData)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }

  changeUserStatus(userId: number) {
    return this.userApiService.changeUserStatus(userId)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }

  getAllUsers(
    pageableParams: PageableParams = {},
    filters: {
      username?: string;
      email?: string;
      fullName?: string;
    } = {}
  ) {
    return this.userApiService.getAllUsers(pageableParams, filters)
      .pipe(
        catchError((error: ApiErrorDetail) => {
          return throwError(() => error.message);
        })
      );
  }
}
